---
title: "Farris et al RNA-seq methods"
output:
  html_document:
    code_folding: show
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: true
    theme: cerulean
    css: liao_style.css
    highlight: pygments
    df_print: kable
  word_document:
    fig_width: 6
    fig_height: 6
    fig_caption: true
always_allow_html: yes
---

```{r setup, include=FALSE}
## To render this page:
## In the farrisSeq folder run:
## rmarkdown::render("farrisSeq.Rmd", "html_document");

suppressPackageStartupMessages(library(knitr));
suppressPackageStartupMessages(library(kableExtra));
suppressPackageStartupMessages(library(dplyr));
suppressPackageStartupMessages(library(plotly));
suppressPackageStartupMessages(library(made4));
suppressPackageStartupMessages(library(matrixStats));

suppressPackageStartupMessages(library(ggplot2));
suppressPackageStartupMessages(library(ggrepel));

suppressPackageStartupMessages(library(jamba));
suppressPackageStartupMessages(library(colorjam));
suppressPackageStartupMessages(library(farrisdata));
knitr::opts_chunk$set(cache=TRUE);

options("kable_styling_position"="left",
  "kable_styling_full_width"=FALSE,
  "kable_styling_bootstrap_options"=c("striped",
    "hover",
    "condensed",
    "responsive")
);
colorSub <- c(
   CA1="orange",
   CA2="orangered3",
   CA3="dodgerblue",
   DG="purple",
   CB="grey75",
   DE="grey30",
   CA1_CB="orange1",
   CA1_DE="darkorange2",
   CA2_CB="orangered1",
   CA2_DE="orangered3",
   CA3_CB="dodgerblue1",
   CA3_DE="dodgerblue3",
   DG_CB="darkorchid1",
   DG_DE="darkorchid3",
   `TRUE`="skyblue",
   `FALSE`="indianred3");
colorSub <- rgb2col(col2rgb(colorSub, alpha=TRUE));
```

This document describes and demonstrates analysis methods used to analyze
mouse hippocampal subregion-specific transcriptome RNA-seq data
(Farris et al manuscript in preparation.)

Raw sequence read data is provided through GEO GSE116343, with RNA-seq
data available at GEO GSE116342. Data used for analysis in R is provided
through an R package `farrisdata`, installed in R with
`devtools::install_github("jmw86069/farrisdata")`.

## Description of Input Data

The input data for this analysis workflow
is derived from Salmon quantitation files, which were already imported and
assembled into relevant data matrices.


### Gene expression data

Salmon quantitation files were imported using the R Bioconductor
package `tximport`, summarized to the gene level, and stored in an
R object `farrisGeneSE`, which is available in the R data package
`farrisdata`, installed using
`devtools::install_github("jmw86069/farrisdata")`.

```{r load_gene_data}
## Load the gene data
library(farrisdata);
library(SummarizedExperiment);
farrisGeneSE;
```

#### farrisGeneSE SummarizedExperiment format

`farrisGeneSE` is a `SummarizedExperiment` object with this content:

* **assays** list containing the numeric matrix with gene rows,
sample columns, with log2 median-normalized counts. Each row is named
using a unique gene symbol derived from the Gencode GTF file. To
access the normalized TPM data:

```{r, eval=FALSE}
assays(farrisGeneSE)[["counts"]]
```

* **assays** also contains the raw Salmon pseudocounts
accessible, using this command: `assays(farrisGeneSE)[["raw_counts"]]`

* **rowData** data.frame with gene rows, using the gene symbol derived from
the Gencode GTF file. 
```{r, gene_table}
as.data.frame(head(rowData(farrisGeneSE), 10)) %>%
   mutate() %>%
   select(probeID, geneSymbol, ends_with("_type"),
      contains("has")) %>%
   kable(escape=FALSE) %>%
   kable_styling()
```

* **colData** annotated data.frame with sample rows, describing the
mouse hippocampal **Region**, **CellType**, **Sample**, and **groupName**.
```{r, sample_table, echo=FALSE}
df2 <- kable_coloring(colData(farrisGeneSE),
      colorSub=colorSub,
      row_color_by="groupName",
      returnType="kable") %>%
   collapse_rows(columns=2:3, valign="middle") %>%
   row_spec(0, background="#DDDDDD")
df2;
```


* **metadata** list describing the experimental design and contrasts used
for statistical comparisons.
    * **design** numeric matrix describing the sample design, suitable
    for use directly by methods from the R Bioconductor `limma` package.
    * **contrasts** numeric matrix describing contrasts, suitable for
    use directly by `limma`.
    * **samples** vector of samples used.
    * **genes** vector of genes used.


### Transcript expression data

Salmon  quantitation files were imported using `tximport` as above,
maintaining individual isoform abundances, then stored in an R object
`farrisTxSE`, which is available in the R data package
`farrisdata`, installed using
`devtools::install_github("jmw86069/farrisdata")`.

```{r load_tx_data}
## Load the gene data
farrisTxSE;
```

#### farrisTxSE format

The `farrisTxSE` object is a named list with content similar to that
described above for gene data:

* **assays** list containing the numeric matrix with gene rows,
sample columns, with log2 median-normalized counts. Each row is named
using a unique gene symbol derived from the Gencode GTF file.
* **rowData** data.frame with gene rows, using the gene symbol derived from
the Gencode GTF file.
```{r, tx_table}
tx_df <- subset(as.data.frame(rowData(farrisTxSE)),
   geneSymbol %in% c("Gria1","Shank2","Ntrk2")) %>%
   select(geneSymbol,
      transcript_id,
      ends_with("_type"),
      contains("has"),
      contains("tpm"));
tx_df <- mixedSortDF(tx_df,
   byCols=match(c("geneSymbol","TxDetectedByTPM"),
     colnames(tx_df))*c(1,-1));
colorSubGene <- c(colorSub,
   group2colors(unique(tx_df$geneSymbol),
      cRange=c(20,30),
      lRange=c(80,90)));
tx_df2 <- kable_coloring(tx_df,
      colorSub=colorSubGene,
      row_color_by="geneSymbol",
      verbose=FALSE,
      returnType="kable") %>%
   collapse_rows(columns=2, valign="middle") %>%
   row_spec(0, background="#DDDDDD")

tx_df2;
```

* **colData** annotated data.frame with sample rows, describing the
mouse hippocampal **Region**, **CellType**, **Sample**, and **groupName**.
This data is identical to the sample table shown for `farrisGeneSE` above.
* **metadata** list describing the experimental design and contrasts used
for statistical comparisons.
    * **design** numeric matrix describing the sample design, suitable
    for use directly by methods from the R Bioconductor `limma` package.
    * **contrasts** numeric matrix describing contrasts, suitable for
    use directly by `limma`.
    * **samples** vector of samples used.
    * **genes** vector of genes used.


## Related analysis parameters

### Sickle read trimming

We used sickle version 1.33 to trim paired-end reads, with parameters
`-q 20 -l 20` to filter reads with quality threshold 20, and requiring
reads at least 20 base length.

### Cutadapt adapter trimming

We used cutadapt version 1.8.1 with parameter `-m 20` and with standard
Illumina Truseq adapter sequence `AGATCGGAAGAGC`.

### STAR sequence alignment

We used STAR version 2.5.1b to align versus Gencode vM12 comprehensive GTF,
with the following parameters, based upon ENCODE recommended RNA-seq
parameters:

```
--outFilterType Normal
--outFilterMultimapNmax 20
--alignSJoverhangMin 8
--alignSJDBoverhangMin 1
--outFilterMismatchNmax 999
--alignIntronMin 20
--alignIntronMax 1000000
--alignMatesGapMax 1000000
--outSAMtype BAM SortedByCoordinate Unsorted
--outWigType bedGraph
--outWigStrand Stranded
--outWigNorm None
--outReadsUnmapped Fastx
--outMultimapperOrder Random
```

### Salmon quantitation

We used salmon version 0.9.1 for transcript quantitation, using
Gencode vM12 comprehensive GTF file, using a index with kmer size 31,
and the cutadapt adapter-trimmed reads, with seqBias, VBO, and gcBias
options enabled.


### featureCounts read counting

We used featureCounts version 1.5.1 to produce per-gene counts, using
Gencode vM12 transcripts flattened per gene.


## Gene expression analysis summary


### BGA plot

```{r bga_plot}
salmonBga1 <- runBga(exprs=assays(farrisGeneSE)$counts,
   signal="counts",
   normMethod="",
   groups=nameVector(colData(farrisGeneSE)$groupName,
      colnames(farrisGeneSE)),
   iSamples=nameVector(colnames(farrisGeneSE)),
   iGenes=rownames(farrisGeneSE),
   noiseFloor=10,
   emptyRowThreshhold=0.6,
   controlColumn="controlGroup",
   fixNames=FALSE,
   verbose=FALSE);
```

```{r bga_plotly}
salmonBga1ly <- bgaPlotly3d(salmonBga1$bgaInfo,
   axes=c(1,2,3),
   colorSub=colorSub,
   useScaledCoords=FALSE,
   drawVectors="none",
   drawSampleLabels=FALSE,
   superGroups=gsub("_.+", "", salmonBga1$bgaInfo$fac),
   ellipseType="none",
   sceneX=0, sceneY=1, sceneZ=1,
   verbose=FALSE);
salmonBga1ly;

```

### Comparison with CA1 published data by Nakayama, Ainsley, Cajigas

First we calculate group medians using gene expression data, to determine
the set of genes detected above an expression threshold of 128 pseudocounts
(log2 = 7).

```{r genes_detected_CA1}
farrisGeneGroupMedians <- rowGroupMeans(assays(farrisGeneSE)[["counts"]],
   groups=colData(farrisGeneSE)$groupName,
   useMedian=TRUE);

# Plot histogram of expression in CA1_CB and CA1_DE
plotPolygonDensity(farrisGeneGroupMedians[,c("CA1_CB","CA1_DE")],
   xlim=c(0,20),
   ylim=c(0,700),
   ablineV=7);

# Detected genes in CA1_CB and CA1_DE
CA1detected <- (farrisGeneGroupMedians[,"CA1_CB"] > 7 &
   farrisGeneGroupMedians[,"CA1_DE"] > 7);
CA1detCBandDE <- rownames(farrisGeneGroupMedians)[CA1detected];
length(CA1detCBandDE);
```

Next we load previously gene lists representing detected CA1 dendrite genes
from published studies from Nakayama et al, Ainsley et al, and
Cajigas et al. These genes are available in the `farrisdata` package,
as described above.

Produce a 4-way Venn diagram showing the overlapping genes from these
studies.

```{r venn_4_way, fig.height=8, fig.width=8}
GeneListL <- list(
  Farris=CA1detCBandDE,
  Cajigas=CajigasGenes,
  Ainsley=AinsleyGenes,
  Nakayama=NakayamaGenes);
GeneListIM <- list2im(GeneListL);
vps <- limma::vennDiagram(GeneListIM,
  circle.col=rainbowJam(4));
```

### Correlation Centered by Compartment, refers to Figure 2

#### Cell Body

Per-sample correlation, centered across all CB samples. First, we
use Salmon normalized pseudocounts, restricted to genes where at
least one group mean value is above log2(7), which is >= 128
normalized pseudocounts.

Then data is centered per **Compartment**, so that CellBody samples
are centered by subtracting the mean CellBody expression per gene,
and Dendrite samples are centered by subtracting the mean Dendrite
expression per gene. This step uses `centerGeneData()` with the
argument `centerGroups`.

Next we prepare a data.frame with color coding to show the
CellType and Compartment values.

```{r sample_corr_cb_prep}
## farrisGeneGroupMedians
## Pull out only cell body samples
iSamples <- colnames(farrisGeneSE);
iSamplesCB <- vigrep("CB", iSamples);
iSamplesDE <- vigrep("DE", iSamples);
iSamplesGrp <- colnames(iMatrix7grp);
iSamplesGrpCB <- vigrep("CB", iSamplesGrp);
iSamplesGrpDE <- vigrep("DE", iSamplesGrp);

corrCutoff <- 7;
genesAboveCutoff <- (rowMaxs(farrisGeneGroupMedians) >= corrCutoff);
genesAboveCutoffCB <- (rowMaxs(farrisGeneGroupMedians[,iSamplesGrpCB]) >= corrCutoff);
genesAboveCutoffDE <- (rowMaxs(farrisGeneGroupMedians[,iSamplesGrpDE]) >= corrCutoff);
genesAboveCutoffBoth <- (genesAboveCutoffCB & genesAboveCutoffDE);

iMatrix7 <- assays(farrisGeneSE[genesAboveCutoff,])[["counts"]];
centerGroups <- gsub("^.+(DE|CB).+$",
   "\\1",
   iSamples);
iMatrix7ctr <- centerGeneData(iMatrix7,
   centerGroups=centerGroups);
iMatrix7ctrCor <- cor(iMatrix7ctr);


## Generate some color bars to annotate the heatmap
colDataColorsRepL <- as.data.frame(colData(farrisGeneSE)[,c("CellType","Compartment")]);
colDataColorsRep <- df2groupColors(colDataColorsRepL,
  colorSub=colorSub);
```

We use `heatmap.3()` to draw the heatmap.

```{r sample_corr_cb, fig.height=8, fig.width=8}
par("family"="Arial", "lend"="square", "ljoin"="mitre");
heatmap.3(iMatrix7ctrCor[iSamplesCB,iSamplesCB],
   margins=c(15,15),
   trace="none",
   col=colorRampPalette(rev(brewer.pal(15,"RdBu")))(51),
   add.expr=box(),
   cexCol=1.5,
   cexRow=1.5,
   keyLabel="Correlation",
   colLensFactor=10,
   distMethod="euclidean",
   hclustMethod="ward",
   doColorDendrograms=TRUE,
   ColSideColors=t(colDataColorsRep[iSamplesCB,2:1]),
   ColSideColorsNote=t(colDataColorsRepL[iSamplesCB,2:1]),
   RowSideColors=(colDataColorsRep[iSamplesCB,2:1]),
   RowSideColorsNote=(colDataColorsRepL[iSamplesCB,2:1]) );
```


#### Cell Body and Dendrite

Correlation showing CA2_CB correlates highest with CA2_DE,
etc. for each CellType.

```{r corr_by_compartment_prep}
## farrisGeneGroupMedians[genesAboveCutoff,]
iMatrix7grp <- farrisGeneGroupMedians[genesAboveCutoffCB,];
centerGroupsGrp <- gsub("^.*(DE|CB).*$",
   "\\1",
   iSamplesGrp);
iMatrix7grpCtr <- centerGeneData(iMatrix7grp,
   centerGroups=centerGroupsGrp);
iMatrix7grpCtrCor <- cor(iMatrix7grpCtr);

## Generate some color bars to annotate the heatmap
colDataColorsGrpL <- data.frame(rbindList(
   strsplit(nameVector(iSamplesGrp), "_")));
colnames(colDataColorsGrpL) <- c("CellType","Compartment");
colDataColorsGrp <- df2groupColors(colDataColorsGrpL,
  colorSub=colorSub);
```

We use `heatmap.3()` to draw the heatmap.

```{r corr_by_compartment, fig.height=8, fig.width=8}
par("family"="Arial", "lend"="square", "ljoin"="mitre");
heatmap.3(iMatrix7grpCtrCor[iSamplesGrp,iSamplesGrp],
   margins=c(15,15),
   trace="none",
   hclCutoff=70,
   cexColSideColorsNote=1.2,
   cexRowSideColorsNote=1.2,
   col=colorRampPalette(rev(brewer.pal(15,"RdBu")))(51),
   add.expr=box(),
   cexCol=1.5,
   cexRow=1.5,
   keyLabel="Correlation",
   colLensFactor=10,
   distMethod="euclidean",
   hclustMethod="ward",
   ColSideColors=t(colDataColorsGrp[iSamplesGrp,2:1]),
   ColSideColorsNote=t(colDataColorsGrpL[iSamplesGrp,2:1]),
   RowSideColors=(colDataColorsGrp[iSamplesGrp,2:1]),
   RowSideColorsNote=(colDataColorsGrpL[iSamplesGrp,2:1]) );
```

#### Splom plot, refers to Figure 2, and Supplemental Figure 3

Scatterplot showing the +1,+1 selection of genes, which
helps define the gene lists in the next section.

```{r splom_CA2, fig.height=8, fig.width=8}
corCols <- c("CA2_DE","CA2_CB");
cutCB <- log2(1.5);
cutDE <- log2(1.5);
splomL <- lapply(nameVector(c("CA1","CA2","CA3","DG")), function(k) {
   corCols <- vigrep(k, colnames(iMatrix7grpCtr));
   df1 <- as.data.frame(iMatrix7grpCtr[,corCols]);
   CBhit <- (abs(df1[,1]) > cutCB) * sign(df1[,1]);
   DEhit <- (abs(df1[,2]) > cutCB) * sign(df1[,2]);
   CBhitN <- paste0(k, "_", "CBhit");
   DEhitN <- paste0(k, "_", "DEhit");
   df1[,CBhitN] <- CBhit;
   df1[,DEhitN] <- DEhit;
   df1[,"GeneName"] <- rownames(df1);
   df1;
});
splomLsub <- lapply(splomL, function(iDF){
   subset(iDF, !iDF[,3] %in% 0 | !iDF[,4] %in% 0)
});

splomDF2 <- rbindList(lapply(splomLsub, function(iDF){
   iDF[,"CellType"] <- gsub("_.+", "", colnames(iDF)[1]);
   colnames(iDF) <- c("CB","DE","CBhit","DEhit","GeneName","CellType");
   iDF;
}));
#splomDF2[,"row"] <- ifelse(splomDF2$CellType %in% c("CA1","CA2"), "CA1", "CA3");
splomDF2[,"CBDEhit"] <- paste0(splomDF2$CBhit,":",splomDF2$DEhit);

## Color-code the splom points
ccColors <- nameVector(rep("grey", 8), unique(splomDF2$CBDEhit));
ccColors[c("1:1","-1:-1")] <- "orangered3";
ccColors[c("-1:1","1:-1")] <- "grey";

## Gene labels
GeneHighlight <- c("Plch2","Rgs14","Necab2","1700024P16Rik");
splomDF2$label <- ifelse(splomDF2$GeneName %in% GeneHighlight,
  splomDF2$GeneName, "");

gg2 <- ggplot(splomDF2,
      aes(x=CB, y=DE, color=CBDEhit, fill=CBDEhit, GeneName=GeneName)) +
   geom_point(shape=21, size=2) +
   geom_vline(xintercept=c(-1,1)*0.585,
      linetype="dashed", color="grey20") +
   geom_hline(yintercept=c(-1,1)*0.585, linetype="dashed", color="grey20") +
   facet_wrap(facets=~CellType) +
   theme_jam() +
   scale_fill_manual(values=ccColors) +
   scale_color_manual(values=makeColorDarker(ccColors)) +
   ggrepel::geom_label_repel(aes(label=label),
      force=8,
      fill="white") +
   theme(legend.position="none");
gg2;
```

Microarray correlation heatmap.

### Neuronal Gene Lists

(Description of the four gene/transcript lists)

Heatmap per-gene.


### Differential gene expression using limma


## Transcript analysis summary

### Definition of "detected transcripts"

We refer to the function `defineDetectedTx()` in the `jampack` R package.
The function takes normalized counts, normalized TPM values, sample
group information, and several cutoff values:

* **cutoffTxPctMax=10** requires a transcript isoform to be at least
10% of the highest isoform present in the same sample.
* **cutoffTxExpr=32** requires a transcript isoform to have at least
32 normalized counts.
* **cutoffTxTPMExpr=2** requires a transcript isoform to have at least
a TPM value of 2.
* All of the above criteria must be met for an isoform to be considered
"detected."

```{r, detectedTxTPM}
refreshFunctions("farrisSalmonWWS");
detectedTxTPML <- defineDetectedTx(
   iMatrixTx=assays(farrisSalmonTxSE)[["counts"]],
   iMatrixTxTPM=assays(farrisSalmonTxSE)[["tpm"]],
   groups=colData(farrisSalmonTxSE)$groupName,
   cutoffTxPctMax=10,
   cutoffTxExpr=32,
   cutoffTxTPMExpr=2,
   tx2geneDF=renameColumn(rowData(farrisSalmonTxSE),
     from=c("geneSymbol","probeID"),
     to=c("gene_name","transcript_id")),
   useMedian=FALSE,
   verbose=FALSE);
```

The code above defined **`r formatInt(length(detectedTxTPML$detectedTx))` detected
transcripts** by the given criteria.


### Transcript type per gene list

e.g. protein_coding, etc.


### Analysis of 3'UTR length

We imported Gencode vM12 comprehensive GTF into R using
R Bioconductor GenomicFeatures package, with which we derived
3'UTR regions, using `makeTxDbFromGFF()` and `threeUTRsByTranscript()`,
respectively.

We plotted 3'UTR lengths as violin plots using only detected transcripts
based upon TPM criteria described elsewhere.


### Translational Efficiency using CAI


### Proximal-Distal 3'UTR Analysis


### Differential transcript isoform analysis limma


### Splicing, Sashimi plots




## Gene Set Enrichment, MultiEnrichMap

### Selection of gene hits

Used DESeq-normalized expression values.


### Hypergeometric enrichment using MSigDB v6.0


#### Preparing MSigDB v6.0 data

Creating gmtT format, refreshing gene symbols in the gmtT data,
and the RNAseq gene hit data to be fully in sync.

Commentary on alt gene symbols, conversion to official mouse
Entrez gene symbols.


#### enrichSimple

Wrapper around standard hypergeometric enrichment tests, with
a custom background of detected genes.


### MultiEnrichMap

Subset pathways for top 15 per source-category

Heatmap of enrichment P-values.

Cnet plot


